### PR TITLE
Added saving Gravatar URL

### LINKED ISSUE #326: Fetch GRAVATAR for User profile using email
**Is your feature request related to a problem? Please describe.**
Currently, the users have an option to set a URL for their profile pictures. Another PR is underway for using AWS for the same, by uploading the images first. However, for a quicker option, it will be good to fetch the profile pictures from GRAVATAR.

Instead of fetching and saving the image, one just needs to grab the URL from GRAVATAR and store it into the profile picture field on the database.

**Describe the solution you'd like**
Fetch GRAVATAR URL using user email and save into database.

### PR DESCRIPTION
**Pull request by bartekpacia**
### Description
When new account is created, gravatar_url is fetched and saved in database.

I'm using [libgravatar](https://pypi.org/project/libgravatar/) to retrieve gravatar url associated with particular email address.

Fixes #326

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Tested manually, with two emails: one with profile picture and one without. First case works as expected, in second case [default gravatar](https://www.gravatar.com/avatar/00000000000000000000000000000000) URL is fetched and saved.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] Update Postman API at /docs folder
- [x] Update Swagger documentation and the exported file at /docs folder
- [x] Update requirements.txt

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by bartekpacia** [comment]
@xprilion I've applied your suggestions, now `gravatar_url` is being assigned to `photo_url`.

File: app/api/dao/user.py [hunk 1]
```diff
@@ -50,6 +51,11 @@ def create_user(data):
         if 'available_to_mentor' in data:
             user.available_to_mentor = data['available_to_mentor']
 
+        gravatar = Gravatar(email)
+        gravatar_url = gravatar.get_image(size=512)
```
Comment (sara-02) [inline app/api/dao/user.py]: This is an unnecessary assignment since we are not using `gravtar_url` again and any future references will be via `user.photo_url`, so we can directly assign `gravatar.get_image(size=512)` to user.photo_url

File: app/api/dao/user.py [hunk 2]
```diff
@@ -50,6 +51,11 @@ def create_user(data):
...
+
+        user.photo_url = gravatar_url
```
Comment (sara-02) [inline app/api/dao/user.py]: IMO, we can put this in a try-catch since we are calling an external API.
WDYT?

**Review by sara-02 — CHANGES_REQUESTED** [review]
Can we also add a small test to make sure that the result fetched from `get_image` function is not an empty string? As in just a test for the `libgravatar`. And raise a warning instead of assert error, this will just be a caution step since the library is not officially supported and we need some way to know if the library starts breaking sometime in future.

**Comment by sara-02** [comment]
I am just a little apprehensive of the unofficially supported library. Rest we are good.

File: app/api/dao/user.py (hunk 2 above)
Comment (bartekpacia) [reply app/api/dao/user.py]: very good idea indeed, fix incoming🛠

**Comment by bartekpacia** [comment]
@Saichethan  I guess that's a good approach. I trust too much :d

File: app/api/dao/user.py (hunk 2 above)
Comment (bartekpacia) [reply app/api/dao/user.py]: Done!

File: app/api/dao/user.py (hunk 1 above)
Comment (bartekpacia) [reply app/api/dao/user.py]: Okay, fixed it:)

File: tests/utils/test_gravatar.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import unittest
+from libgravatar import Gravatar
+
+class TestGravatar(unittest.TestCase):
+
+    def test_works(self):
+        valid_photo_url = "http://www.gravatar.com/avatar/1c968a2c8e98e65b481339b636907d32?size=512"
+
+        gravatar = Gravatar("barpac02@gmail.com")
```
Comment (sara-02) [inline tests/utils/test_gravatar.py]: Let us use `test@test.test` or something random like `test@test.com` :)

**Review by sara-02 — CHANGES_REQUESTED** [review]

File: tests/utils/test_gravatar.py (hunk 3 above)
Comment (bartekpacia) [reply tests/utils/test_gravatar.py]: huh, can't disagree

Comment (bartekpacia) [reply tests/utils/test_gravatar.py]: Fixed 🛠

File: tests/utils/test_gravatar.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import unittest
+
```
Comment (xprilion) [inline tests/utils/test_gravatar.py:2]: Do we need this extra line break?

File: tests/utils/test_gravatar.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+from libgravatar import Gravatar
+
+
+class TestGravatar(unittest.TestCase):
+
+    def test_works(self):
+        valid_photo_url = "http://www.gravatar.com/avatar/b642b4217b34b1e8d3bd915fc65c4452?size=512"
+
```
Comment (xprilion) [inline tests/utils/test_gravatar.py:10]: Do we need this line break?

File: tests/utils/test_gravatar.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        gravatar = Gravatar("test@test.com")
+        photo_url = gravatar.get_image(size=512)
+        print(photo_url)
```
Comment (xprilion) [inline tests/utils/test_gravatar.py]: Can you please remove this print?

File: app/api/dao/user.py [hunk 7]
```diff
@@ -50,6 +51,12 @@ def create_user(data):
         if 'available_to_mentor' in data:
             user.available_to_mentor = data['available_to_mentor']
 
+        try:
+            gravatar = Gravatar(email)
+            user.photo_url = gravatar.get_image(size=512)
+        except:
+            return messages.ERROR_WHILE_CALLING_GRAVATAR, 500
+
```
Comment (xprilion) [inline app/api/dao/user.py]: We do not need this happening while user creation! Pulling a GRAVATAR during user creation (without verification) can have bad consequences. This entire feature should work only after the user is verified!

File: app/api/dao/user.py [hunk 8]
```diff
@@ -200,7 +207,7 @@ def update_user_profile(user_id, data):
                 user.location = data['location']
             else:
                 user.location = None
-            
+
```
Comment (xprilion) [inline app/api/dao/user.py]: Do we need this line break?

**Review by xprilion — CHANGES_REQUESTED** [review]
Hi @bartekpacia, thanks for your work! However, I noticed that you had put this feature during user creation, which is not intended. It should work at two places instead - 

1. When the user is verified, at that moment
2. When the user is already verified and uses some option to change his/her details.

If we run it during user creation without verification, there are chances of undesireable content getting displayed in the app.

**Comment by xprilion** [comment]
Also, you might want to change the PR destination to `gci-dev` branch! 

(Not sure how many other PRs missed this)

**Comment by bartekpacia** [comment]
Okay @xprilion , thanks for advices! Fixes icoming ASAP

I'm still quite new to server-side Python and Flask overall🙄

PS I didn't do it as a part of GCI, that's why I'm targeting `develop`. Other mentors usually ask for doing so.

File: tests/utils/test_gravatar.py (hunk 4 above)
Comment (bartekpacia) [reply tests/utils/test_gravatar.py:2]: Those 2 line breaks are recommended by PEP8.

Comment (bartekpacia) [reply tests/utils/test_gravatar.py:2]: Shall I remove them?

File: tests/utils/test_gravatar.py (hunk 5 above)
Comment (bartekpacia) [reply tests/utils/test_gravatar.py:10]: All those line breaks are recommended by PEP8.

File: app/api/dao/user.py (hunk 8 above)
Comment (bartekpacia) [reply app/api/dao/user.py]: That line break was already there. It just had 3 `tab`s, and I (or rather PyCharm) removed it so it conforms to the rest of the codebase.
But because we are told only to include necessary changes, I added those 3 `tab`s back.

**Comment by bartekpacia** [comment]
@xprilion Okay, I changed behavior of this feature. Now gravatar is fetched and saved in following cases:
* User's account is verified successfully (user clicks on the link in the email)
* User update his account, but leaves photo_url field blank. So basically when `photo_url==null` in `PUT /user`

I also tested above behavior manually - works fine. I'll add unit tests in the near future.

**Comment by sara-02** [comment]
I say we let the line breaks be because they are PEP8 standards, 2 line breaks between import and class definition, and at least one between a class definition and function definitions.

**Comment by xprilion** [comment]
Hey @bartekpacia, thanks for the updates. I shall test them in the morning and let you know! :)

**Comment by bartekpacia** [comment]
@xprilion Thanks, I'm waiting to hear from you:)

**Comment by vj-codes** [comment]
Closing as issue is not assigned to the contributor
